tests/EntryPoints/Specials/SpecialApproverCategoriesTest.php

```diff
@@ -79,19 +66,26 @@ public function testAddApproverAction() {
 		);
 	}
 
-	public function testAddDeleteCategoryAction() {
-		$testUser = $this->getTestSysop()->getUser();
-		$actionParams = [
-			[ 'action' => 'add', 'username' => 'TestUser1', 'category' => 'TestCategory' ],
-			[ 'action' => 'delete', 'username' => 'TestUser1', 'category' => 'TestCategory' ]
-		];
+	public function testAddDeleteCategoryAction(): void {
+		$username = self::getTestUser()->getUser()->getName();
 
-		foreach ( $actionParams as $params ) {
-			$this->executeSpecialPage( $this->createRequestContext( $testUser, $params, true ) );
-		}
+		$this->getPageOutput(
+			request: [
+				'action' => 'add',
+				'username' => $username,
+				'category' => 'TestCategory'
+			]
+		);
+
+		$this->getPageOutput(
+			request: [
+				'action' => 'delete',
+				'username' => $username,
+				'category' => 'TestCategory'
+			]
+		);
 
-		$output = $this->executeSpecialPage( $this->createRequestContext( $testUser ) );
-		$this->assertStringNotContainsString( 'TestCategory', $output, 'Category should be deleted' );
+		$this->assertStringNotContainsString( 'TestCategory', $this->getPageOutput(), 'Category should be deleted' );
```

Comment:
@waximabbax I left the test behavior alone, but this is not a good test yet

Here are some ways the code can still be broken:
* `add` does not do anything
* `delete` removes everything (rather than just what it should)
* the test uses the wrong actions names, so the page does nothing, and the assertion still passes. Have to be careful with negative assertions.

If you can go in the production code and delete lines, flip conditions, or mess up the logic in some other way, and the tests still pass, it means you did not test everything. Testing everything isn't needed, but it's good to test the type of things that are likely to go wrong, either during initial implementation or during refactorings.

Reply:
Perhaps testing Add and Delete Action separately makes more sense. 




**Add** adds the Category that is gonna be deleted. If we don't add the category, we can't confirm if it was deleted or not. 

Actually, I wanted to check if Add and Delete are working through one test. I should asserted to make sure that Add works. This PR ensures that the test doesn't pass if the code is broken.

https://github.com/ProfessionalWiki/PageApprovals/pull/55